test/e2e/storage/testsuites/topology.go

```diff
@@ -0,0 +1,323 @@
+
+		l.resource.sc = dDriver.GetDynamicProvisionStorageClass(l.config, pattern.FsType)
+		framework.ExpectNotEqual(l.resource.sc, nil, "driver failed to provide a StorageClass")
+
+		l.resource.sc.VolumeBindingMode = &pattern.BindingMode
+		// Force all PVs to be provisioned in 1 topology
+		t.setAllowedTopologies(l.topology, l.resource.sc)
+
+		claimSize := dDriver.GetClaimSize()
+		l.resource.pvc = framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+			ClaimSize:        claimSize,
+			StorageClassName: &(l.resource.sc.Name),
+		}, l.config.Framework.Namespace.Name)
+
+		l.intreeOps, l.migratedOps = getMigrationVolumeOpCounts(f.ClientSet, dInfo.InTreePluginName)
+		return l
+	}
+
+	cleanup := func(l topologyTest) {
+		t.cleanupResources(cs, &l)
+		if l.testCleanup != nil {
+			l.testCleanup()
+		}
+
+		validateMigrationVolumeOpCounts(f.ClientSet, dInfo.InTreePluginName, l.intreeOps, l.migratedOps)
+	}
+
+	ginkgo.It("should provision a volume and schedule a pod with AllowedTopologies", func() {
+		l := init()
+		defer func() {
+			cleanup(l)
+		}()
+
+		t.createResources(cs, &l, nil)
+
+		err = e2epod.WaitForPodRunningInNamespace(cs, l.pod)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Verifying pod scheduled to correct node")
+		pod, err := cs.CoreV1().Pods(l.pod.Namespace).Get(l.pod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		node, err := cs.CoreV1().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		t.verifyNodeTopology(node, l.topology)
+	})
+
+	ginkgo.It("should fail to schedule a pod which has topologies that conflict with AllowedTopologies", func() {
+		l := init()
+		defer func() {
+			cleanup(l)
+		}()
+
+		// Set pod nodeSelector to conflicting topology with AllowedTopologies
+		exprs := []v1.NodeSelectorRequirement{}
+		for k, v := range l.topology {
+			exprs = append(exprs, v1.NodeSelectorRequirement{
+				Key:      k,
+				Operator: v1.NodeSelectorOpNotIn,
+				Values:   []string{v},
+			})
+		}
+
+		affinity := &v1.Affinity{
+			NodeAffinity: &v1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+					NodeSelectorTerms: []v1.NodeSelectorTerm{
+						{
+							MatchExpressions: exprs,
+						},
+					},
+				},
+			},
+		}
+		t.createResources(cs, &l, affinity)
+
+		// Wait for pod to fail scheduling
+		// With delayed binding, the scheduler errors before provisioning
+		// With immediate binding, the volume gets provisioned but cannot be scheduled
+		err = e2epod.WaitForPodNameUnschedulableInNamespace(cs, l.pod.Name, l.pod.Namespace)
+		framework.ExpectNoError(err)
+	})
+}
+
+// getCurrentTopologies() goes through all Nodes and returns up to maxCount unique driver topologies
+func (t *topologyTestSuite) getCurrentTopologies(cs clientset.Interface, keys []string, maxCount int) ([]topology, error) {
+	nodes := framework.GetReadySchedulableNodesOrDie(cs)
+
+	topos := []topology{}
+
+	// TODO: scale?
```

Comment:
For a 5000 node cluster in a single test, it's not useful to search through all nodes. But maybe the amount of time it takes here is not a concern in practice.

Reply:
But searching through all nodes is only the worst case. In most cases # topologies is much less than # nodes, and nodes are probably evenly distributed across topologies.

Reply:
yes, I'm saying it's inefficient for large single-zone clusters because the test will look for 2 topologies, but needs to iterate over all 5000 nodes to determine that there isn't enough. anyway i'm not sure it's actually going to matter in practice. we'll see if the large scale tests have problems after this